Constellation Dev Docs (blog-1803) (#8855)
cloudflare/cloudflare-docs: content/constellation/get-started/first-constellation-worker.md

```bash
$ wget https://github.com/microsoft/onnxjs-demo/raw/master/docs/squeezenet1_1.onnx
$ npx wrangler constellation model upload "image-classifier" "squeezenet11" squeezenet1_1.onnx
$ npx wrangler constellation model list "image-classifier"

┌──────────────────────────────────────┬──────────────────────────────────────┬──────────────┐
│ id                                   │ project_id                           │ name         │
├──────────────────────────────────────┼──────────────────────────────────────┼──────────────┤
│ 297f3cda-5e55-33c0-8ffe-224876a76a39 │ 2193053a-af0a-40a6-b757-00fa73908ef6 │ squeezenet11 │
└──────────────────────────────────────┴──────────────────────────────────────┴──────────────┘
```

Take note of the id field as this will be the model id.

## Download Imagenet Classes

The SqueezeNet model was trained on top of the [Imagenet](https://www.image-net.org/) dataset. Let's download the the list of 1000 image classes that it was trained for.

```bash
$ mkdir src
$ wget -O src/imagenet.ts \
  https://raw.githubusercontent.com/microsoft/onnxjs-demo/master/src/data/imagenet.ts
```

> ## Install some modules
> 
> We're going to need [pngjs](https://github.com/pngjs/pngjs), our PNG decoder, and [string-to-stream](https://github.com/feross/string-to-stream).

Reviewer: ```suggestion
## Install modules

Install [pngjs](https://github.com/pngjs/pngjs), a PNG decoder, and [string-to-stream](https://github.com/feross/string-to-stream) before you begin coding your project:
```

    const sumOfExp = resultArray
        .map((resultItem) => Math.exp(resultItem - largestNumber))
        .reduce((prevNumber, currentNumber) => prevNumber + currentNumber);
    return resultArray.map((resultValue) => {
        return Math.exp(resultValue - largestNumber) / sumOfExp;
    });
}

/* Get the top n classes from ImagetNet */

export function topClasses(classProbabilities: any, n = 5) {
    const probabilities = ArrayBuffer.isView(classProbabilities)
        ? Array.prototype.slice.call(classProbabilities)
        : classProbabilities;

    const sorted = probabilities
        .map((prob: any, index: number) => [prob, index])
        .sort((a: Array<number>, b: Array<number>) => {
            return a[0] == b[0] ? 0 : a[0] > b[0] ? -1 : 1;
        });

    const top = sorted.slice(0, n).map((probIndex: Array<number>) => {
        const iClass = imagenetClasses[probIndex[1]];
        return {
            id: iClass[0],
            index: parseInt(probIndex[1].toString(), 10),
            name: iClass[1].replace(/_/g, " "),
            probability: probIndex[0],
        };
    });

    return top;
}

export interface Env {
    CLASSIFIER: any;
}

```

> ## Test it

Reviewer: ```suggestion
## Test your project
```

Here are some test 224x244 PNG images you can use for tests:

```bash
$ wget https://imagedelivery.net/WPOeHKUnTTahhk4F5twuvg/8b78a6fb-44ac-4a97-121b-fb8f47f1e000/public -O cat.png
$ wget https://imagedelivery.net/WPOeHKUnTTahhk4F5twuvg/05c265ae-d3c0-4114-208b-a2d7709cc100/public -O house.png
$ wget https://imagedelivery.net/WPOeHKUnTTahhk4F5twuvg/4152ee23-f9af-4b21-a636-600e33883400/public -O mountain.png
```

### Run wrangler dev

You can start a local server while developing it, for testing.

```bash
$ npx wrangler dev
⬣ Listening at http://0.0.0.0:8787
```

### Run the test images

Now the moment of truth, lets upload the test images to our image classifier.

```bash
$ curl http://0.0.0.0:8787 -F file=@cat.png
{"id":"n02124075","index":285,"name":"Egyptian cat","probability":0.5356272459030151}

$ curl http://0.0.0.0:8787 -F file=@house.png
{"id":"n03028079","index":497,"name":"church","probability":0.5730999112129211}

$ curl http://0.0.0.0:8787 -F file=@mountain.png
{"id":"n09246464","index":972,"name":"cliff","probability":0.37886714935302734}
```

This is it, your image classifier is working. Run it through other 224x244 PNG images of your own and check the results.

> ## Publish it
> 
> When you're ready, you can deploy your Worker to its permanent location.
> 
> ```bash
> $ npx wrangler publish
> ```
> 
> You can see this demo online at https://ai.cloudflare.com/demos/image-classifier

Reviewer: ```suggestion
## Publish your projct

When you are ready, deploy your Worker:

```bash
$ npx wrangler publish
```

[Review this demo online](https://ai.cloudflare.com/demos/image-classifier).
```

---
pcx_content_type: reference
title: First Constellation Worker
meta:
    title: First Constellation Worker
weight: 1
---

# Make your first Constellation Worker

In this example, we will build an Image Classification app powered by a Constellation inference engine and the [SqueezeNet 1.1](https://github.com/onnx/models/blob/main/vision/classification/squeezenet/README.md) ONNX model. SqueezeNet is a small CNN which achieves AlexNet level accuracy on ImageNet with 50x fewer parameters.

> ## Setup project

Reviewer: ```suggestion
## Configure your project
```

> First, let's [create](/constellation/platform/wrangler/#manage-projects) the new project:

Reviewer: ```suggestion
Generate your new project by running [`create`](/constellation/platform/wrangler/#manage-projects). Then run `list` to review th details of your newly created `image-classifier` project:
```
